File: sdks/python/chainbridge/http.py
"""HTTP transport for the ChainBridge SDK (sync + async)."""

from __future__ import annotations

import time
from typing import Any, Dict, Mapping, Optional

import httpx

from .errors import (
    AuthenticationError,
    ChainBridgeError,
    NetworkError,
    NotFoundError,
    RateLimitError,
    ValidationError,
)

DEFAULT_BASE_URL = "https://api.chainbridge.io"
DEFAULT_TIMEOUT = 30.0
USER_AGENT = "chainbridge-sdk-python/0.1.0"


def build_headers(api_key: Optional[str], bearer_token: Optional[str]) -> Dict[str, str]:
    headers: Dict[str, str] = {
        "Content-Type": "application/json",
        "Accept": "application/json",
        "User-Agent": USER_AGENT,
    }
    if api_key:
        headers["X-API-Key"] = api_key
    if bearer_token:
        headers["Authorization"] = f"Bearer {bearer_token}"
    return headers


def _decode_error(status: int, payload: Any) -> ChainBridgeError:
    code = "UNKNOWN_ERROR"
    message = ""
    if isinstance(payload, dict):
        if isinstance(payload.get("error"), dict):
            code = payload["error"].get("code", code)
            message = payload["error"].get("message", "")
        message = message or payload.get("detail") or payload.get("message", "")
    if status == 401 or status == 403:
        return AuthenticationError(message or "Unauthorized")
    if status == 404:
        return NotFoundError(code if code != "UNKNOWN_ERROR" else "NOT_FOUND", message or "Not found")
    if status == 429:
        return RateLimitError(message or "Rate limit exceeded")
    if 400 <= status < 500:
        return ValidationError(code if code != "UNKNOWN_ERROR" else "VALIDATION_ERROR", message or "Bad request")
    return ChainBridgeError(code, message or f"HTTP {status}", status=status, details=payload)


def _unwrap(response: httpx.Response) -> Any:
    if response.status_code == 204 or not response.content:
        return None
    try:
        payload = response.json()
    except ValueError:
        payload = response.text
    if not response.is_success:
        err = _decode_error(response.status_code, payload)
        if isinstance(err, RateLimitError):
            retry_after = response.headers.get("Retry-After")
            if retry_after and retry_after.isdigit():
                err.retry_after_seconds = int(retry_after)
        raise err
    if isinstance(payload, dict) and "success" in payload:
        if payload.get("success") is False:
            err_obj = payload.get("error") or {}
            raise ChainBridgeError(
                err_obj.get("code", "UNKNOWN_ERROR"),
                err_obj.get("message", ""),
                status=response.status_code,
            )
        return payload.get("data")
    return payload


class HttpClient:
    """Synchronous HTTP transport built on httpx with retry + envelope unwrap."""

    def __init__(
        self,
        base_url: str = DEFAULT_BASE_URL,
        api_key: Optional[str] = None,
        bearer_token: Optional[str] = None,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = 3,
        backoff: float = 0.5,
        transport: Optional[httpx.BaseTransport] = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._bearer_token = bearer_token
        self._max_retries = max_retries
        self._backoff = backoff
        self._client = httpx.Client(
            base_url=self.base_url,
            timeout=timeout,
            headers=build_headers(api_key, bearer_token),
            transport=transport,
        )

    def set_bearer_token(self, token: Optional[str]) -> None:
        self._bearer_token = token
        if token is None:
            self._client.headers.pop("Authorization", None)
        else:
            self._client.headers["Authorization"] = f"Bearer {token}"

    def request(
        self,
        method: str,
        path: str,
        *,
        json: Any = None,
        params: Optional[Mapping[str, Any]] = None,
    ) -> Any:
        last_exc: Optional[Exception] = None
        for attempt in range(1, self._max_retries + 1):
            try:
                response = self._client.request(method, path, json=json, params=params)
                return _unwrap(response)
            except (httpx.TransportError, httpx.TimeoutException) as exc:
                last_exc = NetworkError(str(exc), cause=exc)
            except (RateLimitError, ChainBridgeError) as exc:
                if isinstance(exc, RateLimitError) or (
                    isinstance(exc, ChainBridgeError) and exc.status and exc.status >= 500
                ):
                    last_exc = exc
                else:
                    raise
            if attempt < self._max_retries:
                time.sleep(self._backoff * (2 ** (attempt - 1)))
        assert last_exc is not None
        raise last_exc

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> "HttpClient":
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()


class AsyncHttpClient:
    """Async counterpart of :class:`HttpClient`."""

    def __init__(
        self,
        base_url: str = DEFAULT_BASE_URL,
        api_key: Optional[str] = None,
        bearer_token: Optional[str] = None,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = 3,
        backoff: float = 0.5,
        transport: Optional[httpx.AsyncBaseTransport] = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self._max_retries = max_retries
        self._backoff = backoff
        self._client = httpx.AsyncClient(
            base_url=self.base_url,
            timeout=timeout,
            headers=build_headers(api_key, bearer_token),
            transport=transport,
        )

    def set_bearer_token(self, token: Optional[str]) -> None:
        if token is None:
            self._client.headers.pop("Authorization", None)
        else:
            self._client.headers["Authorization"] = f"Bearer {token}"

    async def request(
        self,
        method: str,
        path: str,
        *,
        json: Any = None,
        params: Optional[Mapping[str, Any]] = None,
    ) -> Any:
        import asyncio

        last_exc: Optional[Exception] = None
        for attempt in range(1, self._max_retries + 1):
            try:
                response = await self._client.request(method, path, json=json, params=params)
                return _unwrap(response)
            except (httpx.TransportError, httpx.TimeoutException) as exc:
                last_exc = NetworkError(str(exc), cause=exc)
            except (RateLimitError, ChainBridgeError) as exc:
                if isinstance(exc, RateLimitError) or (
                    isinstance(exc, ChainBridgeError) and exc.status and exc.status >= 500
                ):
                    last_exc = exc
                else:
                    raise
            if attempt < self._max_retries:
                await asyncio.sleep(self._backoff * (2 ** (attempt - 1)))
        assert last_exc is not None
        raise last_exc

    async def aclose(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> "AsyncHttpClient":
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.aclose()

File: sdks/typescript/src/http.ts
import fetchPolyfill from "cross-fetch";
import {
  AuthenticationError,
  ChainBridgeError,
  NetworkError,
  NotFoundError,
  RateLimitError,
  ValidationError,
} from "./errors";
import type { ApiEnvelope } from "./types";

export interface HttpClientOptions {
  baseUrl: string;
  apiKey?: string;
  bearerToken?: string;
  timeoutMs?: number;
  fetch?: typeof fetch;
  defaultHeaders?: Record<string, string>;
  retry?: {
    maxAttempts: number;
    backoffMs: number;
  };
}

const DEFAULT_TIMEOUT_MS = 30_000;

export class HttpClient {
  private readonly baseUrl: string;
  private readonly headers: Record<string, string>;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;
  private readonly retry: { maxAttempts: number; backoffMs: number };

  constructor(opts: HttpClientOptions) {
    this.baseUrl = opts.baseUrl.replace(/\/$/, "");
    this.timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.fetchImpl = (opts.fetch ?? globalThis.fetch ?? fetchPolyfill) as typeof fetch;
    this.retry = opts.retry ?? { maxAttempts: 3, backoffMs: 500 };

    this.headers = {
      "Content-Type": "application/json",
      Accept: "application/json",
      "User-Agent": "chainbridge-sdk-js/0.1.0",
      ...(opts.defaultHeaders ?? {}),
    };
    if (opts.apiKey) this.headers["X-API-Key"] = opts.apiKey;
    if (opts.bearerToken) this.headers["Authorization"] = `Bearer ${opts.bearerToken}`;
  }

  setBearerToken(token: string | null): void {
    if (token) this.headers["Authorization"] = `Bearer ${token}`;
    else delete this.headers["Authorization"];
  }

  async request<T>(
    method: string,
    path: string,
    body?: unknown,
    query?: Record<string, string | number | boolean | undefined>,
  ): Promise<T> {
    const url = this.buildUrl(path, query);
    const init: RequestInit = {
      method,
      headers: this.headers,
      body: body !== undefined ? JSON.stringify(body) : undefined,
    };

    let lastError: unknown;
    for (let attempt = 1; attempt <= this.retry.maxAttempts; attempt++) {
      try {
        return await this.executeOnce<T>(url, init);
      } catch (err) {
        lastError = err;
        if (!this.shouldRetry(err) || attempt === this.retry.maxAttempts) throw err;
        const delay = this.retry.backoffMs * 2 ** (attempt - 1);
        await new Promise((r) => setTimeout(r, delay));
      }
    }
    throw lastError;
  }

  private async executeOnce<T>(url: string, init: RequestInit): Promise<T> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    let res: Response;
    try {
      res = await this.fetchImpl(url, { ...init, signal: controller.signal });
    } catch (err) {
      throw new NetworkError(
        err instanceof Error ? err.message : "Network request failed",
        err,
      );
    } finally {
      clearTimeout(timer);
    }

    const text = await res.text();
    let parsed: ApiEnvelope<T> | { detail?: string } | T | null = null;
    if (text) {
      try {
        parsed = JSON.parse(text);
      } catch {
        if (!res.ok) {
          throw new ChainBridgeError("INTERNAL_ERROR", text || res.statusText, {
            status: res.status,
          });
        }
      }
    }

    if (!res.ok) this.throwApiError(res, parsed);

    if (parsed && typeof parsed === "object" && "success" in parsed) {
      const env = parsed as ApiEnvelope<T>;
      if (env.success === false && env.error) {
        throw new ChainBridgeError(env.error.code, env.error.message, { status: res.status });
      }
      return env.data as T;
    }
    return parsed as T;
  }

  private throwApiError(res: Response, parsed: unknown): never {
    const code = extractCode(parsed) ?? mapStatusToCode(res.status);
    const message = extractMessage(parsed) ?? res.statusText;

    if (res.status === 401 || res.status === 403) throw new AuthenticationError(message);
    if (res.status === 404) throw new NotFoundError(code, message);
    if (res.status === 429) {
      const retryAfter = res.headers.get("Retry-After");
      throw new RateLimitError(message, retryAfter ? Number(retryAfter) : undefined);
    }
    if (res.status >= 400 && res.status < 500) throw new ValidationError(code, message);
    throw new ChainBridgeError(code, message, { status: res.status, details: parsed });
  }

  private shouldRetry(err: unknown): boolean {
    if (err instanceof NetworkError) return true;
    if (err instanceof RateLimitError) return true;
    if (err instanceof ChainBridgeError && err.status && err.status >= 500) return true;
    return false;
  }

  private buildUrl(path: string, query?: Record<string, string | number | boolean | undefined>): string {
    const url = new URL(path.startsWith("/") ? path : `/${path}`, `${this.baseUrl}/`);
    if (query) {
      for (const [k, v] of Object.entries(query)) {
        if (v !== undefined && v !== null) url.searchParams.set(k, String(v));
      }
    }
    return url.toString();
  }
}

function extractCode(parsed: unknown): string | undefined {
  if (parsed && typeof parsed === "object" && "error" in parsed) {
    const err = (parsed as { error?: { code?: string } }).error;
    if (err?.code) return err.code;
  }
  return undefined;
}

function extractMessage(parsed: unknown): string | undefined {
  if (!parsed || typeof parsed !== "object") return undefined;
  const obj = parsed as { error?: { message?: string }; detail?: string; message?: string };
  return obj.error?.message ?? obj.detail ?? obj.message;
}

function mapStatusToCode(status: number): string {
  if (status === 400) return "VALIDATION_ERROR";
  if (status === 401 || status === 403) return "UNAUTHORIZED";
  if (status === 404) return "NOT_FOUND";
  if (status === 429) return "RATE_LIMIT";
  if (status >= 500) return "INTERNAL_ERROR";
  return "UNKNOWN_ERROR";
}
